### ISSUE TITLE
Change _log convention for probability functions

### ISSUE THREAD

**Issue by bob-carpenter** [stan-dev/stan#622]
1.  How the function naming goes in C++ to pick it up:  now end in `_log`.  

2.  How the user calls the function in stan: now ends in `_log`

I suggest we use `_prob` for (1) and drop the suffix for (2).  We can deprecate the `_log` usage as a function.  If we make this change before functions roll out, we don't have to deprecate anything about (1) because it's all behind the scenes.

**Comment by syclik** [comment]
I suggest that we do this for Stan 3 or don't do it at all.

Bob, do you still think we want to rename all our functions `_logprob`? I'm happy to put in the work for Stan 3 if we think this is the right naming convention.

**Comment by syclik** [comment]
I've tagged it for the Stan 3 milestone for now so we can find it.

**Comment by bob-carpenter** [comment]
If we do it, it has to be a major update because it's a whammy
of a backward compatibility breaker. 

In retrospect, I don't think it's worth it to change.

- Bob

**Comment by betanalpha** [comment]
I disagree - the _log_log densities are absolutely atrocious.
_log is a terrible vestige and this is one our chance to fix it.

**Comment by betanalpha** [comment]
Besides, we also talked about kill “~” in favor of
explicit density accumulators, such as

log_post << normal(y, mu, sigma);

There are a few such big changes in the language
that could also be changed at the same time.

**Comment by jgabry** [comment]
So under the proposed name change it would be poisson_logprob
and poisson_log_logprob?

An entirely different approach to changing the names could be to add an
optional argument to the density functions, e.g.

poission(y, lambda, "log"),

which is more aesthetically pleasing (at least to me) and pretty intuitive.

I like the idea of removing the usage of "~" as it doesn't make sense,
except for a tad bit of superficial sense perhaps. It would make it a
bit more confusing for Bugs/Jags users, but maybe by the time Stan 3 comes
out there won't be too many of those anymore. (Except all the discrete
parameter folks, of course.)

**Comment by andrewgelman** [comment]
I like removing “~”

Also, yes, including “log” as an argument is esthetically pleasing but there is a concern in that we will almost always want to be computing log probabilities, almost never will we actually want to be computing unlogged probabilities.

**Comment by jgabry** [comment]
On Tue, Oct 6, 2015 at 9:14 PM, Andrew Gelman <notifications@github.com>
wrote:

>
> Also, yes, including “log” as an argument is esthetically pleasing but
> there is a concern in that we will almost always want to be computing log
> probabilities, almost never will we actually want to be computing unlogged
> probabilities.


That's a good point and having to specify the log all the time would be
annoying. Although we're already using poisson_log(y, lambda) for this,
which isn't less typing than poisson(y, lambda, "log"). Either way we need
to put the "log" somewhere.

**Comment by andrewgelman** [comment]
Yes, I agree.  The point is that poisson_log is the default, whereas it might be awkward if it’s poisson(…), where non-logged is the default.

**Comment by betanalpha** [comment]
I like _ld as an economical reference to “log density”.

**Comment by bob-carpenter** [comment]
Actually, we don't need to put "log" anywhere.

I was lazy in using it in the language so that what I
do is allow any function whose name ends in "_log" to
be used on the right-hand-side of a sampling statement
without the log, 

   y ~ normal(mu, sigma);

translates to:

   increment_log_prob(normal_log(y, mu, sigma));

So "normal" on the right side of ~ and "normal_log"
as a function.  But if we're going to redo this, we can start
from scratch.

This will affect user-defined functions, which follow the "_log"
convention as implemented and documented --- mainly that's just
to play into the ~ syntax.  If we get rid of that, there's then
nothing special about user-defined densities.

If we're going to something like a log prob accumulator
strategy, we could do something like:

  log_post << normal(y, mu, sigma);

or we could even do what Andrew asked for ages ago and write

  log_post << normal(y | mu, sigma);

The problem from a readability point of view is that "log_post <<"
is long and written up front.  We could alternatively use a suffix
notation

  normal(y | mu, sigma) >> log_post;

or something shorter in the prefix department;  we really
only need

  << normal(y | mu, sigma);

unless we're thinking of different accumulators (say for "prior"
and "likelihood" [quotes to try to nip further discussion of
what these mean in the bud]) --- the problem with doing that is
we can't tease apart the blocks and execute them separately.
Sort of the same problem we have with not being able to compute
Markov blankets for discrete sampling now.

So a model might look like this:

  model {
    << cauchy(alpha | 0, 2.5);
    << cauchy(beta | 0, 2.5);
    << lognormal(sigma | 0, 2);
    << normal(y | alpha + beta * x, sigma);
  }

instead of

  model {
    alpha ~ cauchy(0, 2.5);
    beta ~ cauchy(0, 2.5);
    sigma ~ lognormal(0, 2);
    y ~ normal(alpha + beta * x, sigma);
  }

Of course, if we introduce the "|" notation, we might want to
restrict where it's used to replace ",", which brings back the
issue of the density functions being "special" in some way.  
We could use a set of names for the ones we define and would then
have to come up with some way to mark user-defined functions
as densities.

Do we do the same as before for truncation?

   << poisson(y | lambda) T[1, ];

Right now, the syntax for user-defined functions works as usual,
with "_log" and "_cdf_log" being used if they're defined for
user-defined functions and being illegal otherwise.

I don't like "log_post" because we're not really saying anything
about this being a posterior --- it can be any old unnormalized
density.

- Bob

**Comment by bob-carpenter** [comment]
The problem with "log density" is that we're often defining
a log joint that contains discrete parameters which just
happen to be fixed as data or computed some other way.  That's
why I went with "log prob" in the first place, thinking along
the lines of Andrew's use of "log probability function" for
both PMFs and PDFs in BDA.  Getting rid of the "function" bit of
it, though, was a huge mistake, as it's too easily confused
with "log probability", wheras it's usually a density or mixed
density and mass function.

- Bob

**Comment by wds15** [comment]
interesting discussion. Just a thought: The world is still full of so many people who rely on some BUGS variant. The current Stan language is very familiar to all of those people, making it easy to switch. This would not be any longer the case when you do such drastic changes (even though they admittedly look nice). So maybe for Stan 4 this is ok, but maybe not for now? When Stan 4 arrives many will have turned to Stan already and the BUGS world has gotten considerably smaller by then...

In short: BUGS similarity is a good thing for many out there!

**Comment by jgabry** [comment]
Would it be possible in the language to introduce the lp accumulator
notation in Stan 3 without disallowing "~" sampling notation until Stan 4,
whenever that is? A soft introduction, so to speak?

**Comment by bob-carpenter** [comment]
Yes, we could deprecate "~" and spit out a warning indicating
how to translate the statement so the warning goes away.

- Bob

**Comment by andrewgelman** [comment]
+1

**Comment by bob-carpenter** [comment]
Won't "poisson_ld" look weird because it's a mass function,
not a density function?  It'd be an easy change if we just
changed the suffix (or even left the old one with a warning).

- Bob

**Comment by andrewgelman** [comment]
Yes, I like these ideas!

**Comment by betanalpha** [comment]
One of the natural extensions of having named accumulators is that it
would be trivial to have more than one, such as prior and likelihood or
base and delta or whatever you want to call them.

The density function vs mass function is indeed a confusing issue that
causes a gigantic amount of pedagogical damage.  I agree that using 
any common name will only add to the confusion as they confound 
canonical objects (probability mass functions, which are densities of 
a discrete probability distribution with respect to the canonical counting 
measure) with non-canonical objects (probability density functions, which 
are densities with respect to an arbitrary base measure).  

That said, they are technically both densities.  Certainly writing
poisson_radon_nikodym_derivative will be too annoying.  :-p
Hence I still prefer log density (or _ld) and we can be more careful to
elaborate that Stan defines a probabiltiy density function with respect
to the continuous Lebesgue measure implied by the parameters.

But really anything that doesn’t conflict with link functions would be great.

**Comment by bob-carpenter** [comment]
> On Oct 8, 2015, at 11:41 AM, Michael Betancourt <notifications@github.com> wrote:
> 
> One of the natural extensions of having named accumulators is that it
> would be trivial to have more than one, such as prior and likelihood or
> base and delta or whatever you want to call them.

Multiple named accumulators are problematic in that it becomes
tricky to compute just one of them.   And when they're in different
blocks, it's hard to specify and hard to share subcomputations.  I
don't know what the right answer is.  Certainly it'd be easy to have
multiple named accumulators as long as we always ran the whole model
block, even if we just threw away parts of it.  Or we could not execute
the increment statements for one block as long as there aren't any
side effects (we're in luck there, with Stan being call-by-value
semantically).

> The density function vs mass function is indeed a confusing issue that
> causes a gigantic amount of pedagogical damage. I agree that using 
> any common name will only add to the confusion as they confound 
> canonical objects (probability mass functions, which are densities of 
> a discrete probability distribution with respect to the canonical counting 
> measure) with non-canonical objects (probability density functions, which 
> are densities with respect to an arbitrary base measure). 
> 
> That said, they are technically both densities. 

Makes sense.  If you want two things to look the same, jump
up a level of mathematics! 

- Bob

**Comment by rtrangucci** [comment]
@bob-carpenter seems to be subsumed under the larger changes we're making the language, no?

**Comment by bob-carpenter** [comment]
yes, and I'll create new issues for this as I go.

**Comment by ariddell** [comment]
@bob-carpenter I was looking for a current spec of the language changes and ended up here. Is there a place one should look for a draft of the spec?

**Comment by TRANTANKHOA** [comment]
Hi everyone,

Is there also a list of functions available in STAN math’s library along with the description of their interfaces? Something like a manual for the library itself, not for STAN.

I’ve seen the index.html in the doc folder but I just want to check if that is all we have for this purpose?

Thanks!
Khoa.

**Comment by bob-carpenter** [comment]
There's the doxygen-generated documentation, but it's not
particularly complete beyond the signatures.

Can we move this discussion somewhere relevant, as it's off-issue
for #622 here.

**Comment by bob-carpenter** [comment]
Here's everything that's coming:

https://github.com/stan-dev/stan/wiki/Stan-3-Density-Notation-and-Increments

I'll deprecate ~ as a final move if everyone wants, but I'm
still skeptical it's a good idea.

**Comment by betanalpha** [comment]
> I'll deprecate ~ as a final move if everyone wants, but I'm
> still skeptical it's a good idea.

We can avoid going all the way to deprecation, but we need
to do _something_.  In the mathematics literature everyone
reads ~ as “is distributed according to” while many practitioners
coming from BUGS and JAGS read is as “is sampled from”.
They are roughly equivalent conceptually but _very_ different
operationally, and it’s the difference in the operational definition
that is causing all of the problems.  Not sure how best to do this
other than having a “Guide for former BUGS and JAGS users”.

**Comment by bob-carpenter** [comment]
We do have that in an appendix and it already says this,
but we can't get people to read more than a couple pages before
writing 200-line Stan programs!

- Bob

**Comment by betanalpha** [comment]
Right, which means we need a new approach if we want to keep the tilde!
